Navigation: Add create new menu button

scruffian opened the pull request:
## What?
This handles the case where the "create new menu" option is displayed as a button, when in the offcanvas editing experiment.

Before:
<img width="284" alt="Screenshot 2022-11-22 at 13 21 17" src="https://user-images.githubusercontent.com/275961/203324347-646967d8-729e-4ca8-a535-76ebe7d9ad0c.png">

After:
<img width="290" alt="Screenshot 2022-11-22 at 13 14 34" src="https://user-images.githubusercontent.com/275961/203323969-3e05704c-d16e-4521-b16b-e39049b3ad1f.png">

## Why?
Because before looks bad!

## How?
I moved the button to a new component. I tried to do some refactoring to avoid us having too much duplicated logic for the condition which determines which state we are in. I tried getting the createNavigationMenu function from the new component but it didn't work in the same way as passing it through the props. We should try to address that in a followup.

## Testing Instructions
1. Delete all classic menus and all wp_navigation menus
2. Enable the offcanvas editing experiment
3. Add a navigation block
4. Confirm that the "Create new menu" button is not alongside the "Menu" heading.

scruffian committed:
Navigation: Add create new menu button

getdave commented on packages/block-library/src/navigation/edit/create-new-menu-button.js:
```diff
@@ -0,0 +1,27 @@
+			__experimentalIsFocusable
+			onClick={ () => {
+				onCreateNew();
+				setIsCreatingMenu( true );
```
It is unusual that we need an internal var for this. We can't pass the state down into the component?

scruffian replied:
I was trying to avoid passing props around. What would be the benefit of sharing state here? I believe it's only used to control the UI.

getdave replied:
We're using internal state that isn't connected to the actual result of the action we're invoking.

That _could_ leave our button in a permanent "loading" state. We're relying on the component being unmounted in order to reset the internal state back to not loading. If this button ends up in a different context to that which you are expecting here it may introduce bugs.

An alternative would be to pass the resolution state down into the component and use that to conditionalise the internal UI.

Currently in this example the function that ends up being called is

```js
const createUntitledEmptyNavigationMenu = () => {
	createNavigationMenu( '' );
};
```

Here the error state isn't being handled inline at all but rather it's deferred to the `useEffect`s in the main component. We should have a `.catch` (that's for another PR).

The point is that `createNavigationMenu` could easily error and the `CreateNewMenuButton` would know nothing about it and could potentially end up loading forever.

Either the component is controlled or it is uncontrolled. I feel like it's currently a bit of both.

getdave commented on packages/block-library/src/navigation/edit/menu-inspector-controls.js:
```diff
@@ -86,6 +90,38 @@ const MenuInspectorControls = ( {
+							/>
+						) }
+					</>
+				) : (
```
Urgh. This branching around offcanvas is getting unwieldy 🤷

getdave commented on packages/block-library/src/navigation/edit/use-more-navigation-menu.js:
```diff
@@ -0,0 +1,15 @@
+import useNavigationMenu from '../use-navigation-menu';
+import useNavigationEntities from '../use-navigation-entities';
+
+const useMoreNavigationMenu = () => {
```
I can't remember if I suggested this on the other PR. The naming of the hook seems to be tied to the UI rather than to the concept of what it represents which is "do we have either block or classic menus".

Also it's singular when the hook is looking for for multiple.

🙏 Please can we consider renaming to reflect the idea of the hook?
```suggestion
const useHasNavigationMenus = () => {
```

getdave reviewed (commented):
Again thank you for looking into this annoying bug! 👏 

This does appear to work as described. It is slightly unusual in visual appearance - it's not really a button or a link.

What I don't understand is why when we have Page List we don't just put "create new" in the 3 dots menu?

#### Experiment enabled

<img width="693" alt="Screen Shot 2022-11-23 at 10 22 49" src="https://user-images.githubusercontent.com/444434/203523134-e14f7f32-532f-4ded-b79b-36a0c3c3f096.png">


#### Experiment disabled

<img width="666" alt="Screen Shot 2022-11-23 at 10 24 17" src="https://user-images.githubusercontent.com/444434/203523434-225805da-f306-4a3e-94a7-ad1a78848971.png">

scruffian commented:
> What I don't understand is why when we have Page List we don't just put "create new" in the 3 dots menu?

I think this is because this counts as a fallback, but maybe that's not a good reason.

scruffian committed:
always put create new menu in the drop down for the experiment

scruffian commented:
Thanks for the review @getdave. I've updated the name of the hook, and I've updated the behaviour so that "Create new menu" is always part of the more menu in the experiment.

getdave commented on packages/block-library/src/navigation/edit/use-has-classic-or-navigation-menus.js:
```diff
@@ -0,0 +1,15 @@
+import useNavigationMenu from '../use-navigation-menu';
+import useNavigationEntities from '../use-navigation-entities';
+
+const useHasClassicOrNavigationMenus = () => {
```
❤️

draganescu commented on packages/block-library/src/navigation/edit/menu-inspector-controls.js:
```diff
@@ -70,6 +72,8 @@ const MenuInspectorControls = ( {
 		: undefined;
 
+	const hasClassicOrNavigationMenus = useHasClassicOrNavigationMenus();
+
```
Given the UI is not the more menu and one button in the tools section, that button can be there all the time. So we don't need to check for this anymore.

draganescu reviewed (changes requested):
I like the UI resulting from this PR. But I think the PR should try less. In the current UI _all the PR has to do_ is:
- remove the behavior that transforms the dropdown into a button
- if inside the experiment.

scruffian committed:
remove unnecessary hook

scruffian commented:
OK third time's a charm? https://github.com/WordPress/gutenberg/pull/46019
